Add alias dispatch key DefaultBackend.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45939 [WIP] Move catchAll to Math
* #45938 Update native_functions.yaml to add DefaultBackend.
* #45719 Support DefaultBackend keyword in native_functions.yaml.
* **#45718 Add alias dispatch key DefaultBackend.**

Differential Revision: [D24165892](https://our.internmc.facebook.com/intern/diff/D24165892)

Commit:
Add alias dispatch key DefaultBackend.

[ghstack-poisoned]

ezyang commented:
Don't forget to add more detailed PR description. This can be copy paste from the design doc, but it helps a lot for external viewers who aren't familiar with the back story.

ezyang commented on test/test_dispatch.py:
```diff
@@ -571,6 +571,44 @@ def test_computed_table_with_ambiguous_autogradother(self):
+AutogradOther: fn_math [math kernel]
+AutogradCPU: fn_autograd [autograd kernel]
+AutogradCUDA: fn_math [math kernel]
+AutogradXLA: fn_math [math kernel]
```
This feels... wrong? Default backend should count as a backend for the purposes of math registration, which means you should get `fn_autograd` here? (So it seems to me like Math shouldn't show up at all)

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -181,39 +181,49 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
+  }
+
   bool is_autograd_key_with_backend_kernel =
     hasKernelForDispatchKeySet(getBackendKeySetFromAutograd(dispatch_key));
```
So it seems to me that this should return true if we have DefaultBackend

Commit:
Update on "Add alias dispatch key DefaultBackend."




[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key DefaultBackend."




[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key DefaultBackend."




[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -181,61 +181,79 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
-    hasKernelForDispatchKeySet(getBackendKeySetFromAutograd(dispatch_key));
-  // 2.1. Use Math kernel if available. For autograd keys, we only use kernel from Math
-  //      when there's no direct registration to its corresponding backend key.
+    hasKernelForDispatchKeySet(getBackendKeySetFromAutograd(dispatch_key).add(DispatchKey::DefaultBackend));
```
With this modification, it occurs to me that is not obvious the "has kernel" is a "ALL" or an "ANY" relation here (looking at the code, it's ANY, but that's not obvious from the name)

ezyang commented on test/test_dispatch.py:
```diff
@@ -571,6 +571,111 @@ def test_computed_table_with_ambiguous_autogradother(self):
+        self.assertExpectedInline(extracted_table, '''\
+CPU: fn_cpu [kernel]
+CUDA: fn_defaultbackend [default backend kernel]
+XLA: fn_defaultbackend [default backend kernel]
```
Marginally better if you add a backend that is part of AutogradOther here.

Commit:
Update on "Add alias dispatch key DefaultBackend."




[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -181,61 +181,79 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
-    hasKernelForDispatchKeySet(getBackendKeySetFromAutograd(dispatch_key));
-  // 2.1. Use Math kernel if available. For autograd keys, we only use kernel from Math
-  //      when there's no direct registration to its corresponding backend key.
+    hasKernelForAnyDispatchKey(getBackendKeySetFromAutograd(dispatch_key).add(DispatchKey::DefaultBackend));
```
I think there might be an issue here - recapping my understanding of what this was doing before:

- if `dispatch_key` is an autograd key, `getBackendKeySetFromAutograd(dispatch_key)` will return the set of backend keys it maps to, otherwise the empty keyset.

- `hasKernelForDispatchKeySet(<keyset>)` returns true if we've registered a kernel for any of the keys in the set

...so the meaning of `is_autograd_key_with_backend_kernel` is, "`dispatch_key` is an autograd key and it maps to at least one backend key we have a kernel for".

So if that's accurate, then I think adding `DefaultBackend` like this changes the logic, since `is_autograd_key_with_backend_kernel` will now be true if we have a kernel registered for any backend key at all, regardless of whether `dispatch_key` is an autograd key or not, right?

ailzhang replied:
@bhosmer Note that we passed `DefaultBackend` to `hasKernelForAnyDispatchKey` instead of its Runtime keys. So `is_autograd_key_with_backend_kernel` will be true when there's a kernel registered to `DefaultBackend` or any backend in `autograd_other_backends` ;)

bhosmer replied:
Ah you're right, I was thinking about the alias set :P but still, say we have a kernel registered to `DefaultBackend` directly. Won't `is_autograd_key_with_backend_kernel` now be true for any value of `dispatch_key` at all? 

I'm not saying this necessarily breaks the logic where `is_autograd_key_with_backend_kernel` is used (I'm assuming it doesn't, since tests pass etc. :D) , but if my understanding is right, I think we should at least rename the variable (or even split it).

ailzhang replied:
> say we have a kernel registered to DefaultBackend directly. Won't is_autograd_key_with_backend_kernel now be true for any value of dispatch_key at all?
yep for any *Autograd* keys this will be true if there's a DefaultBackend. But I feel this kinda fits the current name `is_autograd_key_with_backend_kernel` where we just think of `DefaultBackend` as direct registration to all backends?

bhosmer replied:
sorry to be dense :) but what I'm not seeing is why this wouldn't be true for all dispatch keys, not just autograd keys. Like say dispatch_key == CPU. Then getBackendKeySetFromAutograd(dispatch_key).add(DispatchKey::DefaultBackend) would be {DispatchKey::DefaultBackend}, right?

The only other thing I can think of is that there are invariants that prevent non-autograd dispatch_key values from ever reaching this code, but I'm not sure I see how those would work... 

Anyway I'm sure I'm just missing something basic. You can set me straight tmr when we chat :)

ezyang replied:
It's guarded downstream by `isIncludedInAlias(dispatch_key, DispatchKey::Math)` and `isIncludedInAlias(dispatch_key, DispatchKey::Autograd)`. Agree that the structure is confusing though.

bhosmer reviewed (commented):
Hey Ailing - gotta jump off this for a bit but wanted to get you the one observation I'm making inline - I could be wrong but if not it might be useful. 

Oh also I see the PR adds a `bak` file, not sure this was intended :) 

Back for more in a bit!

bhosmer reviewed (approved):
Accepting to unblock, we can follow up tomorrow on the inline thread - if it's clear what I'm getting wrong about the current code please go ahead and land!

Also for tomorrow I wanna pitch you on a potential simplification (not high pri, just something to think about for later on) - the basic idea is to split computeDispatchTableEntryWithDebug into a handful of simpler functions that only handle particular kinds of dispatch keys. It occurred to me as I read thru this that we might be making things more complicated than they need to be by concentrating all the smarts in that single function - since we actually know at the call sites what kind of dispatch key we're sending in.

Commit:
Update on "Add alias dispatch key DefaultBackend."


Differential Revision: [D24165892](https://our.internmc.facebook.com/intern/diff/D24165892)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key DefaultBackend."


Differential Revision: [D24165892](https://our.internmc.facebook.com/intern/diff/D24165892)

[ghstack-poisoned]

ailzhang commented:
Updated the https://github.com/pytorch/pytorch/issues/44680 with use cases and motivation, will send a followup PR to update the README in native/ folder.
